src/app/fdctl/run/tiles/fd_repair.c

```diff
@@ -0,0 +1,656 @@
+
+  for( ulong i=0UL; i<dest_cnt; i++ ) {
+    fd_repair_peer_addr_t repair_peer = {
+      .addr = in_dests[i].ip4_addr,
+      .port = in_dests[i].udp_port,
+    };
+   
+    fd_repair_add_active_peer( ctx->repair, &repair_peer, in_dests[i].pubkey );
+  }
+}
+
+static inline void
+handle_new_repair_requests( fd_repair_tile_ctx_t * ctx,
+                            uchar const *          buf,
+                            ulong                  buf_sz ) {
+
+  fd_repair_request_t const * repair_reqs = (fd_repair_request_t const *) buf;
+  ulong repair_req_cnt = buf_sz;
+
+  for( ulong i=0; i<repair_req_cnt; i++ ) {
+    fd_repair_request_t const * repair_req = &repair_reqs[i];
+    int rc = 0;
+    switch(repair_req->type) {
+      case FD_REPAIR_REQ_TYPE_NEED_WINDOW_INDEX: {
+        rc = fd_repair_need_window_index( ctx->repair, repair_req->slot, repair_req->shred_index );
+        break;
+      }
+      case FD_REPAIR_REQ_TYPE_NEED_HIGHEST_WINDOW_INDEX: {
+        rc = fd_repair_need_highest_window_index( ctx->repair, repair_req->slot, repair_req->shred_index );
+        break;
+      }
+      case FD_REPAIR_REQ_TYPE_NEED_ORPHAN: {
+        rc = fd_repair_need_orphan( ctx->repair, repair_req->slot );
+        break;
+      }
+    }
+    
+    if( rc < 0 ) {
+      FD_LOG_DEBUG(( "failed to issue repair request" ));
+    }
+  }
+
+}
+
+static inline void
+handle_new_stake_weights( fd_repair_tile_ctx_t * ctx ) {
+  ulong stakes_cnt = ctx->stake_ci->scratch->staked_cnt;
+
+  if( stakes_cnt >= MAX_REPAIR_PEERS ) {
+    FD_LOG_ERR(( "Cluster nodes had %lu stake weights, which was more than the max of %lu", stakes_cnt, MAX_REPAIR_PEERS ));
+  }
+
+  fd_stake_weight_t const * in_stake_weights = ctx->stake_ci->stake_weight;
+  fd_repair_set_stake_weights( ctx->repair, in_stake_weights, stakes_cnt );
+}
+
+
+static void 
+repair_send_packet( uchar const *                 msg, 
+                    size_t                        msglen, 
+                    fd_gossip_peer_addr_t const * addr, 
+                    void * arg ) {
+  ulong tsorig = fd_frag_meta_ts_comp( fd_tickcount() );
+  send_packet( arg, addr->addr, fd_ushort_bswap( addr->port ), msg, msglen, tsorig );
+}
+
+static void
+repair_shred_deliver( fd_shred_t const *            shred,
+                      ulong                         shred_sz, 
+                      fd_repair_peer_addr_t const * from FD_PARAM_UNUSED, 
+                      fd_pubkey_t const *           id FD_PARAM_UNUSED, 
+                      void *                        arg ) {
+  fd_repair_tile_ctx_t * ctx = (fd_repair_tile_ctx_t *)arg;
+  
+  fd_shred_t * out_shred = fd_chunk_to_laddr( ctx->store_out_mem, ctx->store_out_chunk );
+  fd_memcpy( out_shred, shred, shred_sz );
+  
+  ulong tspub = fd_frag_meta_ts_comp( fd_tickcount() );
+  ulong sig = 0UL;
+  fd_mux_publish( ctx->mux, sig, ctx->store_out_chunk, shred_sz, 0UL, 0UL, tspub );
```

Comment:
How many shreds can be requested to deliver per turn of the repair_continue? You might only have 1 flow control credit here (due to burst being set to 1 below), so if repair_continue tries to deliver two shreds this could corrupt the shreds.

Reply:
The deliver function only allows for 1 shred to be delivered